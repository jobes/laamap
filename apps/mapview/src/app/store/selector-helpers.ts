export function barDefinition(val: {
  minShownValue: number;
  maxShownValue: number;
  alertLower: number;
  alertUpper: number;
  cautionLower: number;
  cautionUpper: number;
  position: { x: number; y: number };
  bgColor: string;
  textColor: string;
}) {
  const absoluteAlertLowerPercentage = Math.min(
    100,
    ((val.alertLower - val.minShownValue) /
      (val.maxShownValue - val.minShownValue)) *
      100,
  );
  const absoluteCautionLowerPercentage = Math.min(
    100,
    ((val.cautionLower - val.minShownValue) /
      (val.maxShownValue - val.minShownValue)) *
      100,
  );
  const absoluteCautionUpperPercentage = Math.min(
    100,
    ((val.cautionUpper - val.minShownValue) /
      (val.maxShownValue - val.minShownValue)) *
      100,
  );
  const absoluteAlertUpperPercentage = Math.min(
    100,
    ((val.alertUpper - val.minShownValue) /
      (val.maxShownValue - val.minShownValue)) *
      100,
  );
  const alertLowerPercentage = absoluteAlertLowerPercentage;
  const cautionLowerPercentage =
    absoluteCautionLowerPercentage - absoluteAlertLowerPercentage;
  const goodPercentage =
    absoluteCautionUpperPercentage - absoluteCautionLowerPercentage;
  const cautionUpperPercentage =
    absoluteAlertUpperPercentage - absoluteCautionUpperPercentage;
  const alertUpperPercentage = 100 - absoluteAlertUpperPercentage;
  return {
    alertLowerPercentage,
    cautionLowerPercentage,
    goodPercentage,
    cautionUpperPercentage,
    alertUpperPercentage,
    position: val.position,
    bgColor: val.bgColor,
    textColor: val.textColor,
  };
}

export function barExtraValues(
  currentValue: number | null,
  settings: {
    alertLower: number;
    alertUpper: number;
    cautionLower: number;
    cautionUpper: number;
    minShownValue: number;
    maxShownValue: number;
    show: boolean;
  },
  instrumentsConnected: boolean,
) {
  if (
    !instrumentsConnected ||
    !settings.show ||
    currentValue === null ||
    currentValue === undefined
  ) {
    return {
      alert: false,
      caution: false,
      currentValue: 0,
      valuePercentage: 0,
      show: false,
    };
  }
  const percentage =
    ((currentValue - settings.minShownValue) /
      (settings.maxShownValue - settings.minShownValue)) *
    100;
  return {
    show: true,
    alert:
      currentValue < settings.alertLower || currentValue > settings.alertUpper,
    caution:
      currentValue < settings.cautionLower ||
      currentValue > settings.cautionUpper,
    currentValue: currentValue,
    valuePercentage: percentage < 0 ? 0 : percentage > 100 ? 100 : percentage,
  };
}
