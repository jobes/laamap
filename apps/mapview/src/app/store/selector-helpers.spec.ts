import { barDefinition, barExtraValues } from './selector-helpers';

describe('Selector helpers', () => {
  describe('barExtraValues', () => {
    it('should map with no caution and alert', () => {
      expect(
        barExtraValues(
          170,
          {
            alertLower: 70,
            alertUpper: 200,
            cautionLower: 150,
            cautionUpper: 180,
            minShownValue: 50,
            maxShownValue: 250,
            show: true,
            compact: false,
          },
          true,
        ),
      ).toMatchSnapshot();
    });

    it('should map with lower caution', () => {
      expect(
        barExtraValues(
          140,
          {
            alertLower: 70,
            alertUpper: 200,
            cautionLower: 150,
            cautionUpper: 180,
            minShownValue: 50,
            maxShownValue: 250,
            show: true,
            compact: false,
          },
          true,
        ),
      ).toMatchSnapshot();
    });

    it('should map with upper caution', () => {
      expect(
        barExtraValues(
          190,
          {
            alertLower: 70,
            alertUpper: 200,
            cautionLower: 150,
            cautionUpper: 180,
            minShownValue: 50,
            maxShownValue: 250,
            show: true,
            compact: false,
          },
          true,
        ),
      ).toMatchSnapshot();
    });

    it('should map with lower alert', () => {
      expect(
        barExtraValues(
          60,
          {
            alertLower: 70,
            alertUpper: 200,
            cautionLower: 150,
            cautionUpper: 180,
            minShownValue: 50,
            maxShownValue: 250,
            show: true,
            compact: false,
          },
          true,
        ),
      ).toMatchSnapshot();
    });

    it('should map with upper alert', () => {
      expect(
        barExtraValues(
          210,
          {
            alertLower: 70,
            alertUpper: 200,
            cautionLower: 150,
            cautionUpper: 180,
            minShownValue: 50,
            maxShownValue: 250,
            show: true,
            compact: false,
          },
          true,
        ),
      ).toMatchSnapshot();
    });

    it('should map when value lower than min bar', () => {
      expect(
        barExtraValues(
          30,
          {
            alertLower: 70,
            alertUpper: 200,
            cautionLower: 150,
            cautionUpper: 180,
            minShownValue: 50,
            maxShownValue: 250,
            show: true,
            compact: false,
          },
          true,
        ),
      ).toMatchSnapshot();
    });

    it('should map when value bigger than max bar', () => {
      expect(
        barExtraValues(
          270,
          {
            alertLower: 70,
            alertUpper: 200,
            cautionLower: 150,
            cautionUpper: 180,
            minShownValue: 50,
            maxShownValue: 250,
            show: true,
            compact: false,
          },
          true,
        ),
      ).toMatchSnapshot();
    });

    it('should be hidden when not connected', () => {
      expect(
        barExtraValues(
          270,
          {
            alertLower: 70,
            alertUpper: 200,
            cautionLower: 150,
            cautionUpper: 180,
            minShownValue: 50,
            maxShownValue: 250,
            show: true,
            compact: false,
          },
          false,
        ),
      ).toMatchSnapshot();
    });

    it('should be hidden when not shown', () => {
      expect(
        barExtraValues(
          270,
          {
            alertLower: 70,
            alertUpper: 200,
            cautionLower: 150,
            cautionUpper: 180,
            minShownValue: 50,
            maxShownValue: 250,
            show: false,
            compact: false,
          },
          true,
        ),
      ).toMatchSnapshot();
    });

    it('should be hidden when value null', () => {
      expect(
        barExtraValues(
          null,
          {
            alertLower: 70,
            alertUpper: 200,
            cautionLower: 150,
            cautionUpper: 180,
            minShownValue: 50,
            maxShownValue: 250,
            show: true,
            compact: false,
          },
          true,
        ),
      ).toMatchSnapshot();
    });
  });

  describe('barDefinition', () => {
    it('should define normal bars', () => {
      expect(
        barDefinition({
          alertLower: 70,
          alertUpper: 130,
          cautionLower: 90,
          cautionUpper: 120,
          bgColor: 'red',
          position: { x: 20, y: 30 },
          textColor: 'black',
          minShownValue: 50,
          maxShownValue: 150,
        }),
      ).toMatchSnapshot();
    });

    it('should define normal bars overflowed alert', () => {
      expect(
        barDefinition({
          alertLower: 70,
          alertUpper: 130,
          cautionLower: 90,
          cautionUpper: 120,
          bgColor: 'red',
          position: { x: 20, y: 30 },
          textColor: 'black',
          minShownValue: 50,
          maxShownValue: 110,
        }),
      ).toMatchSnapshot();
    });

    it('should define normal bars overflowed caution', () => {
      expect(
        barDefinition({
          alertLower: 70,
          alertUpper: 130,
          cautionLower: 90,
          cautionUpper: 120,
          bgColor: 'red',
          position: { x: 20, y: 30 },
          textColor: 'black',
          minShownValue: 50,
          maxShownValue: 125,
        }),
      ).toMatchSnapshot();
    });
  });
});
