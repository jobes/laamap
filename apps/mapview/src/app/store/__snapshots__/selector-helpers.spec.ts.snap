// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Selector helpers barDefinition should define normal bars 1`] = `
{
  "alertLowerPercentage": 20,
  "alertUpperPercentage": 20,
  "bgColor": "red",
  "cautionLowerPercentage": 20,
  "cautionUpperPercentage": 10,
  "goodPercentage": 30,
  "position": {
    "x": 20,
    "y": 30,
  },
  "textColor": "black",
}
`;

exports[`Selector helpers barDefinition should define normal bars overflowed alert 1`] = `
{
  "alertLowerPercentage": 33.33333333333333,
  "alertUpperPercentage": 0,
  "bgColor": "red",
  "cautionLowerPercentage": 33.33333333333333,
  "cautionUpperPercentage": 0,
  "goodPercentage": 33.33333333333334,
  "position": {
    "x": 20,
    "y": 30,
  },
  "textColor": "black",
}
`;

exports[`Selector helpers barDefinition should define normal bars overflowed caution 1`] = `
{
  "alertLowerPercentage": 26.666666666666668,
  "alertUpperPercentage": 0,
  "bgColor": "red",
  "cautionLowerPercentage": 26.666666666666668,
  "cautionUpperPercentage": 6.666666666666671,
  "goodPercentage": 39.99999999999999,
  "position": {
    "x": 20,
    "y": 30,
  },
  "textColor": "black",
}
`;

exports[`Selector helpers barExtraValues should be hidden when not connected 1`] = `
{
  "alert": false,
  "caution": false,
  "currentValue": 0,
  "show": false,
  "valuePercentage": 0,
}
`;

exports[`Selector helpers barExtraValues should be hidden when not shown 1`] = `
{
  "alert": false,
  "caution": false,
  "currentValue": 0,
  "show": false,
  "valuePercentage": 0,
}
`;

exports[`Selector helpers barExtraValues should be hidden when value null 1`] = `
{
  "alert": false,
  "caution": false,
  "currentValue": 0,
  "show": false,
  "valuePercentage": 0,
}
`;

exports[`Selector helpers barExtraValues should map when value bigger than max bar 1`] = `
{
  "alert": true,
  "caution": true,
  "currentValue": 270,
  "show": true,
  "valuePercentage": 100,
}
`;

exports[`Selector helpers barExtraValues should map when value lower than min bar 1`] = `
{
  "alert": true,
  "caution": true,
  "currentValue": 30,
  "show": true,
  "valuePercentage": 0,
}
`;

exports[`Selector helpers barExtraValues should map with lower alert 1`] = `
{
  "alert": true,
  "caution": true,
  "currentValue": 60,
  "show": true,
  "valuePercentage": 5,
}
`;

exports[`Selector helpers barExtraValues should map with lower caution 1`] = `
{
  "alert": false,
  "caution": true,
  "currentValue": 140,
  "show": true,
  "valuePercentage": 45,
}
`;

exports[`Selector helpers barExtraValues should map with no caution and alert 1`] = `
{
  "alert": false,
  "caution": false,
  "currentValue": 170,
  "show": true,
  "valuePercentage": 60,
}
`;

exports[`Selector helpers barExtraValues should map with upper alert 1`] = `
{
  "alert": true,
  "caution": true,
  "currentValue": 210,
  "show": true,
  "valuePercentage": 80,
}
`;

exports[`Selector helpers barExtraValues should map with upper caution 1`] = `
{
  "alert": false,
  "caution": true,
  "currentValue": 190,
  "show": true,
  "valuePercentage": 70,
}
`;
